docs/style-guide.md

```diff
@@ -0,0 +1,439 @@
+
+Many of the practices here are collected from existing sources (which should be cited) and where applicable there may be some additional justification for including that practice in this document and why it is categorized as it is. In general where a practice is sensible, expected to be achievable, and has no obvious downside, it will be marked as required.
+
+Note however that none of this document is meant to trump common sense, if if you're in a situation where it would be better to violate a **required** practice, then it just means we should have a discussion about it. There will almost certainly be many cases where we need to update this doc and even go back to existing code to update it in the light of new information. However exceptions should always be commented clearly for the next maintainer.
+
+## Sources
+
+* [Learn Rust](https://doc.rust-lang.org/)
+  * The canonical source for Rust information, but it doesn't offer advice on all aspects of code design.
+* [Introduction - Learning Rust With Entirely Too Many Linked Lists](https://rust-unofficial.github.io/too-many-lists/)
+  * This is brilliantly written and very educational about how you get into a Rust mindset.
+* [Elegant Library APIs in Rust](https://deterministic.space/elegant-apis-in-rust.html)
+  * This has a lot of good points and is well worth a read, but be warned that some of the details about implementation are somehwat out of date (2017). It has a video too.
+* [Good Practices for Writing Rust Libraries](https://pascalhertleif.de/artikel/good-practices-for-writing-rust-libraries/)
+  * Shorter and more focussed on the act of coding the libraries, and less about API design. Also potentially out of date in places.
+* [Strategies for Returning References in Rust](http://bryce.fisher-fleig.org/blog/strategies-for-returning-references-in-rust/index.html)
+  * Though I don't believe we should be doing this, it's still an interesting read.
+
+## Other Useful Links
+
+* Write and run Rust snippets: https://play.rust-lang.org
+  * You can save snippets in permantent links and incluce them as working examples in docs.
+* Write code and see what it compiles to: https://rust.godbolt.org
+  * Note that you need to enable compiler optimizations via `-C opt-level=3` if you are looking to meaningfully compare two code snippets.
+
+# Naming Conventions
+
+### No Special Naming :: required
+
+Follow the naming advice in [Naming - Rust API Guidelines](https://rust-lang.github.io/api-guidelines/naming.html) at all times, and use [Rust RFC 0430](https://github.com/rust-lang/rfcs/blob/master/text/0430-finalizing-naming-conventions.md) for any ambiguous issues (e.g. Crate naming).
+
+Based on [this comment](https://github.com/rust-lang/api-guidelines/issues/29#issuecomment-342745898) I think the advice for Crate names should be:
```

Comment:
"I think" sounds like a personal stance, not a project guideline.

Reply:
Fair enough. This is another way this doc has morphed from "David's personal thoughts" to "project guidelines". I'll reword.